### ISSUE TITLE
WaveformModel from .onnx

### ISSUE THREAD

**Issue by MarcoEsposito890** [seisbench/seisbench#205]
Hello, I have a custom machine learning model in .onnx format that I would like to import in seisbench to annotate obspy streams. 

I'm having issues converting the model to pytorch (strangely, I get slightly different predictions from the two formats). Would it be possible to do so or would the best way to do this be converting it to pytorch first?

I'm mainly interested in inheriting the annotations methods

**Comment by yetinam** [comment]
Hi @MarcoEsposito890 ,

I personally have no experience with ONNX, so I'm afraid I can't help you. In principle, internally the actual prediction step happens [here](https://github.com/seisbench/seisbench/blob/fdf8351a2347f86c56ed28a1c7fe47f2bc275304/seisbench/models/base.py#L1813). So if you can provide anything that is callable with a pytorch tensor as input and outputs a python tensor you should be able to fiddle it in there. Alternatively, you could build a class inheriting from `WaveformModel` and overwrite exactly the `_predict_buffer` function.
Let me know this goes, I'm curious to see if we could add an ONNX interface.

**Comment by MarcoEsposito890** [comment]
> Hi @MarcoEsposito890 ,
> 
> I personally have no experience with ONNX, so I'm afraid I can't help you. In principle, internally the actual prediction step happens [here](https://github.com/seisbench/seisbench/blob/fdf8351a2347f86c56ed28a1c7fe47f2bc275304/seisbench/models/base.py#L1813). So if you can provide anything that is callable with a pytorch tensor as input and outputs a python tensor you should be able to fiddle it in there. Alternatively, you could build a class inheriting from `WaveformModel` and overwrite exactly the `_predict_buffer` function. Let me know this goes, I'm curious to see if we could add an ONNX interface.

Hi, thanks for the tips. I will work on it starting from there.

My first (brutal) attempt was to create a class the following way, in which I invoke the onnx runtime and pass it the model path. It works to make predictions on standalone windows, but it crashes when using annotations. I will try to overwrite the _predict_buffer function and see if I can make it work

'
    """
    testing onnx
    """
    def __init__(
        self,
        in_channels=3,
        classes=3,
        phases="NPS",
        sampling_rate=100,
        norm="std",
        **kwargs,
    ):
        # Skip super call in favour of super-super class
        WaveformModel.__init__(
            self,
            in_samples=401,
            output_type="array",
            pred_sample=(0, 401),
            labels=phases,
            sampling_rate=sampling_rate,
            **kwargs,
        )

        self.onnx_model_path = onnx_model_path
        ort_sess = ort.InferenceSession(onnx_model_path)

    def forward(self, x, logits=False):
        
        x = ort_sess.run(None, {'input': x })
        x = x[0]

        return torch.from_numpy(x)`

**Comment by MarcoEsposito890** [comment]
This is a not very elegant solution that works. The issues I had before when running the model on a stream were mainly due to mismatched dimensions between the _predict_buffer arguments and the output of the prediction with the onnx runtime.

Invoking the onnx runtime inside the forward method of a class subclassing nn.module is probably not the best way to go, but the fastest to run some tests using your annotate method. I'll reopen the issue if I have time to update this with an onnx-only implementation.

`from seisbench.models.base import Conv1dSame, WaveformModel

class testModel(WaveformModel):
    """
    testing onnx
    """


    def __init__(
        self,
        in_channels=3,
        classes=3,
        phases="PNS",
        sampling_rate=100,
        norm="std",
        **kwargs,
    ):
        # Skip super call in favour of super-super class
        WaveformModel.__init__(
            self,
            in_samples=401,
            output_type="array",
            pred_sample=(0, 401),
            labels=phases,
            sampling_rate=sampling_rate,
            **kwargs,
        )

        self.onnx_model_path = onnx_model_path
        ort_sess = ort.InferenceSession(onnx_model_path)

    def forward(self, x, logits=False):
        
        #x = x/max(x)
        x = ort_sess.run(None, {'input': x })
        x = x[0]
        #print(x.shape)

        return torch.from_numpy(x)

    def _predict_buffer(self, buffer):
        """
        Batches model inputs, runs prediction, and unbatches output

        :param buffer: List of inputs to the model
        :return: Unpacked predictions
        """
        fragments = np.stack(buffer, axis=0).squeeze() ## onnx model takes batched input
        fragments = np.array(fragments, dtype = np.float32)
        preds = self(fragments)
        preds = self._recursive_torch_to_numpy(preds)
        reshaped_preds = [np.swapaxes(pred,0,1) for pred in self._recursive_slice_pred(preds)]
        print(reshaped_preds[0].shape)
        return reshaped_preds`

![image](https://github.com/seisbench/seisbench/assets/48209182/220fff1c-8514-4403-ab04-cd0f927e711e)

**Comment by MarcoEsposito890** [comment]
Side question: the annotate function for point prediction models cannot be used to predict over a sliding window, am I correct? By looking at the metadata saved in the _cut_fragments_array function I see that the start times of each fragment is a multiple of the model input lenght, so I'm assuming that point prediction models annotations are applied to the trace divided in fixed length windows

**Comment by yetinam** [comment]
Yes, the `overlap` defines exactly this. `overlap=0` will cause non-overlapping windows, higher values give the number of samples to consecutive windows should overlap in.

**Comment by MarcoEsposito890** [comment]
Ok, thank you very much!
